amethyst_network/src/network_socket.rs

```diff
@@ -70,89 +45,90 @@ where
 {
     /// The list of filters applied on the events received.
     pub filters: Vec<Box<dyn NetFilter<E>>>,
-
-    tx: Sender<InternalSocketEvent<E>>,
-    rx: Receiver<RawEvent>,
+    // sender on which you can queue packets to send to some endpoint.
+    transport_sender: Sender<InternalSocketEvent<E>>,
+    // receiver from which you can read received packets.
+    transport_receiver: Receiver<Packet>,
+    config: ServerConfig,
 }
 
 impl<E> NetSocketSystem<E>
 where
     E: Serialize + PartialEq + Send + 'static,
 {
     /// Creates a `NetSocketSystem` and binds the Socket on the ip and port added in parameters.
-    pub fn new(addr: SocketAddr, filters: Vec<Box<dyn NetFilter<E>>>) -> Result<Self, Error> {
-        if addr.port() < 1024 {
+    pub fn new(config: ServerConfig, filters: Vec<Box<dyn NetFilter<E>>>) -> Result<Self> {
+        if config.udp_recv_addr.port() < 1024 {
             // Just warning the user here, just in case they want to use the root port.
             warn!("Using a port below 1024, this will require root permission and should not be done.");
         }
 
-        let mut socket = UdpSocket::bind(addr, NetworkConfig::default())
-            .map_err(|x| Error::new(ErrorKind::Other, x.to_string()))?;
+        let server = Host::run(config.clone())?;
 
-        socket.set_nonblocking(true).map_err(|_| {
-            Error::new(
-                ErrorKind::Other,
-                "Unable to set `UdpSocket` to non-blocking mode",
-            )
-        })?;
+        let server_sender = NetSocketSystem::<E>::start_sending(&server);
```

Comment:
this block isn't clear to me why do you send a reference in one call but not in the other? Does a clone happen on the other side if so I would like this to be moved up to this piece of code.

Reply:
`start_sending` takes the reference and `start_receiving` will eventually own the instance. It is just a way to play whit the borrowing.

Reply:
Yeah, I see now this confuses me, to be honest, I find it unclear what is being borrowed and for how long. Since you are passing a channel handle down do you think that this function could take that handle instead?

Reply:
Agree, I have to experiment with this further, I tried some things a few seconds ago but it is a little difficult to borrow those Handles.

Reply:
done, I had to move `SendHandler` into an Arc. Although for a `Receiver` it is more difficult to move it around threads since it does not implement `Sync`. For this, I had to move it into a `Mutex`